Source/SIMPLib/ITK/itkFijiConfigurationFileReader.hpp

```diff
@@ -0,0 +1,127 @@
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+* USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*
+* The code contained herein was partially funded by the followig contracts:
+*    United States Air Force Prime Contract FA8650-07-D-5800
+*    United States Air Force Prime Contract FA8650-10-D-5210
+*    United States Prime Contract Navy N00173-07-C-2068
+*
+* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
+
+#pragma once
+
+#include <QtCore/QObject>
+#include <QtCore/QString>
+#include <QtCore/QPointF>
+#include <QtCore/QFileInfo>
+#include <QtCore/QDir>
+
+#include "itkParseTileConfiguration.h"
+
+namespace itk
+{
+struct FijiImageTileData
+{
+  QString filePath;
+  QPointF coords;
+  int row;
+  int col;
+};
+
+typedef std::vector<FijiImageTileData> FijiRowData;
+typedef std::vector<FijiRowData> FijiFileData;
+
+class FijiConfigurationFileReader : public QObject
+{
+  public:
+    FijiFileData parseFijiConfigFile(const QString &configFilePath)
+    {
+      m_ErrorCondition = 0;
+      itk::FijiFileData fileData;
+
+      itk::TileLayout2D stageTiles = itk::ParseTileConfiguration2D(configFilePath.toStdString());
+      for (int i = 0; i < stageTiles.size(); i++)
+      {
+        auto tile2D = stageTiles[i];
+        FijiRowData rowData;
+        for (int j = 0; j < tile2D.size(); j++)
+        {
+          auto tile = tile2D[j];
+
+          QString imageFilePath = QString::fromStdString(tile.FileName);
+          QFileInfo fi(imageFilePath);
+          if (fi.isRelative())
+          {
+            QFileInfo configFi(configFilePath);
+            imageFilePath = configFi.path() + QDir::separator() + imageFilePath;
+          }
+
+          itk::FijiImageTileData imageTileData;
+
+          imageTileData.filePath = imageFilePath;
+
+          uint32_t coordsSize = tile.Position.GetPointDimension();
+          if (coordsSize != 2)
+          {
+            m_ErrorMessage = QObject::tr("The dimension size of all tiles in the fiji config file must be equal to 2.");
+            m_ErrorCondition = -2004;
+            return itk::FijiFileData();
+          }
+
+          imageTileData.coords = QPointF(tile.Position[0], tile.Position[1]);
+          imageTileData.row = i;
+          imageTileData.col = j;
+
+          rowData.push_back(imageTileData);
+        }
+
+        fileData.push_back(rowData);
+      }
+
+      return fileData;
```

Comment:
Unless a method or class is templated, constexpr, or inline, try to avoid putting method definitions in the header.  I'm aware that this is done in other places, so I'm not going to say it needs to be changed.  Just try to avoid it in the future.

This seems to be specific enough it won't matter much, but the reason to avoid it is to avoid forcing recompiles for every file that even indirectly includes this if even a single line of a method definition changes.  The reason templates, constexpr, and inline are excused are because they require definitions where they are declared for the sake of compiling.  Constexpr and inline are both pretty forceful runtime optimizations and require consideration when using in part for these same reasons.  Considering constexpr throws a fit with strings, two of the three reasons are already ruled out.  Inline is only useful for small methods like getErrorCode() and getErrorMessage() when they are called in tight loops or other time-critical places where runtime is slowed down due to the function call overhead.  Do note that this, like constexpr, is handled at compile time and will slow that down slightly in exchange for the runtime performance.

Reply:
This file I cannot speak to. Idk if @joeykleingers has anything to chime in and say? regardless, I don't think it has any bearing on the mutli-selection widget nature of the PR.